### DISCUSSION TITLE
RP2350 A and B variant

### DISCUSSION THREAD

**Discussion by elral** [earlephilhower/arduino-pico#2371] [Ideas]
The new RP2350 is available in variant `A` and `B`. The `A` variant has the same number of pins as the RP2040 where the `B` variant has 40 pins (and at least the analog pins on another pins).
For now there is one generic variant for the RP2350. I am wondering if it would make sense to mark this variant as `A` and setup another one for variant `B`.

Yesterday I moved our project from platformIO to the Arduino IDE to test it with the RP2350. For what I could test everything was working fine (except  AnaologIn as I am using an PGA2350, but will test it with a Pico2 again).
Really great job!! Thanks a lot for your effort!

**Comment by earlephilhower** [comment]
They're both the same die, so it should just be a matter of adding the new `GPIO_HI_*` registers when they're used explicitly (only IRQ masking in SPI/Ethernet IIRC) and adjusting the `digital/analogWrite/Read` to support >31 GPIOs.

**Reply by earlephilhower** [reply]
Actually, scratch that :disappointed: .  They do have a strapping and wiring difference which makes them incompatible.  May need completely new `libpico` rebuild just for it.  For example, ADC will be wired up to upper GPIOs which will differ -A vs -B

![image](https://github.com/user-attachments/assets/74736100-a6cc-4d87-adeb-6787d4802cb5)
![image](https://github.com/user-attachments/assets/73334a8a-3499-4964-bb25-c5e3ba798b3b)

**Reply by PontusO** [reply]
Maybe move the low level analog stuff into the bsp instead. Seems over the top to have a new lib simply for the adc.

**Reply by earlephilhower** [reply]
TBH I'm not sure that this is the only code/user-visible difference, it's just the first one I noticed.  The SDK sometimes pulls in lots of dependencies so it may not be possible, but it's worth a look.

In the worst case end users won't know about it, there will just be one more pass in `tools/libpico` and things will "just work."

**Reply by elral** [reply]
I set up a new board this morning to test it out. After copying the generic one I just extended the pins and defined the new analog pins.
For what I can see everything was working, but for analog pins just tested A0/1 on Pin 40/41 and not all 8 analog pins.
